### DISCUSSION TITLE
Build for macOS + m1 chips

### DISCUSSION THREAD

**Discussion by jrabary** [sail-sg/envpool#200] [General]
Hi all,
I'm not sure whether macOS is already supported. I'm trying to build on my laptop with m1 chips and got the following error

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
clang++: warning: argument unused during compilation: '-msse' [-Wunused-command-line-argument]
clang++: warning: argument unused during compilation: '-msse2' [-Wunused-command-line-argument]
clang++: warning: argument unused during compilation: '-mmmx' [-Wunused-command-line-argument]
In file included from envpool/box2d/box2d_envpool.cc:15:
In file included from ./envpool/box2d/bipedal_walker.h:21:
In file included from ./envpool/core/async_envpool.h:30:
In file included from ./envpool/core/envpool.h:23:
In file included from ./envpool/core/env_spec.h:23:
./envpool/core/dict.h:58:23: warning: string literal operator templates are a GNU extension [-Wgnu-string-literal-operator-template]
inline constexpr auto operator""_() {  // NOLINT
                      ^
In file included from envpool/box2d/box2d_envpool.cc:15:
In file included from ./envpool/box2d/bipedal_walker.h:21:
./envpool/core/async_envpool.h:110:9: error: unknown type name 'cpu_set_t'
        cpu_set_t cpuset;
        ^
./envpool/core/async_envpool.h:114:70: error: use of undeclared identifier 'cpu_set_t'; did you mean 'cpuset'?
        pthread_setaffinity_np(workers_[tid].native_handle(), sizeof(cpu_set_t),
                                                                     ^~~~~~~~~
                                                                     cpuset
./envpool/core/async_envpool.h:110:19: note: 'cpuset' declared here
        cpu_set_t cpuset;
                  ^
In file included from envpool/box2d/box2d_envpool.cc:18:
In file included from ./envpool/core/py_envpool.h:33:
./envpool/core/xla.h:20:10: fatal error: 'cuda_runtime_api.h' file not found
#include <cuda_runtime_api.h>
         ^~~~~~~~~~~~~~~~~~~~
1 warning and 3 errors generated.
Error in child process '/usr/bin/xcrun'. 1
Target //:setup failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

So two errors related to intel chips and Nvidia cuda.

Any ideas to fix this ?

**Comment by Trinkle23897** [comment]
Thanks for reporting this. Closing the loop here for anyone who lands on this thread later.

This was a real gap back in 2022: the source build still assumed Linux-only CPU-affinity APIs and CUDA/XLA headers.

Current status:
- EnvPool now supports macOS in current releases, including PyPI wheels, so the recommended path on Apple Silicon is just `pip install envpool`.
- The `cpu_set_t` / `pthread_setaffinity_np` path is now gated to Linux only.
- XLA is still unavailable on macOS, so JAX/XLA workflows are not supported there yet.

If you do need to build from source on macOS, please follow the current build doc:
- `xcode-select --install`
- `brew install go openjdk@17 swig qt@5`
- `export PATH="$(brew --prefix openjdk@17)/bin:$PATH"`
- `export BAZEL_RULES_QT_DIR="$(brew --prefix qt@5)"`

So the short answer is: the specific M1 build errors in this thread should no longer apply to current EnvPool, but XLA on macOS is still unsupported.

Refs:
- https://pypi.org/project/envpool/
- https://envpool.readthedocs.io/en/latest/content/build.html
